File: frontend/src/modules/robots/config/types/type2-tinvest.ts
import type {
    RobotCostsConfig,
    RobotRiskConfig,
    RobotStrategyName,
    RobotStrategyParams,
} from '@/types/robot'

export type MoexRefreshSchedule = {
    every_minutes: number
    only_trading_hours: boolean
    daily_at_msk: string | null
}

export type MoexPipelineFilter = {
    type: string
    min?: number
    max_percent?: number
    min_percent?: number
    period?: number
    eq?: string
    direction?: string
    max_steps?: number
    min_ratio?: number
    list?: string[] | null
}

export type MoexHistoricalScreeningConfig = {
    enabled: boolean
    source: 'moex' | 'tinvest'
    board: string
    universe: string
    fixed_tickers: string[]
    interval: string
    lookback_days: number
    filters: MoexPipelineFilter[]
    refresh: MoexRefreshSchedule
}

export type MoexPaperSelectionConfig = {
    enabled: boolean
    input: string
    fixed_tickers: string[]
    mode: 'ALL' | 'ANY'
    filters: MoexPipelineFilter[]
    refresh: MoexRefreshSchedule
}

export type MoexSignalGenerationConfig = {
    strategy: RobotStrategyName
    params: RobotStrategyParams
    data_source: string
    update_interval_seconds: number
    indicator_update_schedule: Record<string, string>
}

/** MOEX trading robot config v3 (`schema_profile=type2_tinvest`). */
export type Type2TinvestConfig = {
    config_version: 3
    schema_profile: 'type2_tinvest'
    instrument_id_type: 'figi'
    broker_type: 'tinvest'
    strategy: RobotStrategyName
    strategy_params: RobotStrategyParams
    allowed_figis: string[]
    universe_mode?: 'fixed' | 'dms_pipeline' | 'tqbr_scan'
    fixed_tickers?: string[]
    universe_refresh_minutes?: number
    update_interval_seconds: number
    indicator_update_schedule: Record<string, string>
    pipeline: {
        mode: 'ALL' | 'ANY'
        filters: MoexPipelineFilter[]
    }
    risk: RobotRiskConfig
    costs: RobotCostsConfig
    historical_screening: MoexHistoricalScreeningConfig
    paper_selection: MoexPaperSelectionConfig
    signal_generation: MoexSignalGenerationConfig
    instrument_map?: Record<string, unknown>
}

File: backend/app/modules/robots/config/profiles/type2_tinvest.py
"""Profile schema: trading robot type=2, broker=tinvest."""

from __future__ import annotations

from typing import Literal

from pydantic import Field

from app.modules.robots.config.costs_moex import MoexCostsConfig
from app.modules.robots.config.risk_moex import MoexRiskConfig
from app.modules.robots.config.v2_schema import TradingRobotConfigV2


class Type2TinvestConfig(TradingRobotConfigV2):
    """Профиль type2_tinvest (v3 identity + v2 pipeline blocks)."""

    config_version: Literal[3] = 3
    schema_profile: Literal["type2_tinvest"] = "type2_tinvest"
    broker_type: Literal["tinvest", "sandbox"] = "tinvest"
    instrument_id_type: Literal["figi"] = "figi"
    risk: MoexRiskConfig = Field(default_factory=MoexRiskConfig)
    costs: MoexCostsConfig = Field(default_factory=MoexCostsConfig)


__all__ = ["Type2TinvestConfig"]